sessions/2018/november/session-18.md

```diff
@@ -0,0 +1,198 @@
+
+[Here][4] documentation about how to integrate BugSnag when using sourcemaps.
+
+
+
+#### Making partial git commits
+
+When I want to add to a specific commit just a part of the code that I have modified, the following command can be used:
+
+$ git add -p
+
+And this will be asking for confirmation to add specific chunks of code (I can split in more chunks if I want)
+
+Example:
+
+```git
+$ git add -p
+diff --git a/build.gradle b/build.gradle
+index ad5c980..87c18ed 100644
+--- a/build.gradle
++++ b/build.gradle
+@@ -31,4 +31,6 @@ dependencies {
+        runtime('org.springframework.boot:spring-boot-devtools')
+        testCompile('org.springframework.boot:spring-boot-starter-test')
+        testCompile('de.flapdoodle.embed:de.flapdoodle.embed.mongo')
++       compile 'com.bugsnag:bugsnag-spring:3.+'
++
+ }
+Stage this hunk [y,n,q,a,d,e,?]? y
+
+diff --git a/src/main/java/co/org/osso/api/CalculatorController.java b/src/main/java/co/org/osso/api/CalculatorController.java
+index 2d2549a..a61e283 100644
+--- a/src/main/java/co/org/osso/api/CalculatorController.java
++++ b/src/main/java/co/org/osso/api/CalculatorController.java
+@@ -1,5 +1,6 @@
+ package co.org.osso.api;
+
++import com.bugsnag.Bugsnag;
+ import org.springframework.web.bind.annotation.RestController;
+ import org.springframework.web.bind.annotation.RequestMapping;
+ import org.springframework.web.bind.annotation.GetMapping;
+Stage this hunk [y,n,q,a,d,j,J,g,/,e,?]? y
+@@ -14,13 +15,16 @@ import java.util.List;
+ public class CalculatorController {
+
+     private final Calculator calculator;
++    private final Bugsnag bugsnag;
+
+-    public CalculatorController(Calculator calculator) {
++    public CalculatorController(Calculator calculator, Bugsnag bugsnag) {
+         this.calculator = calculator;
++        this.bugsnag = bugsnag;
+     }
+
+     @GetMapping("/fibonacci/{number}")
+     public List<BigInteger> getFibonacci(@PathVariable("number") Integer number){
++        bugsnag.notify(new RuntimeException("Test error"));
+         return calculator.getFibonacci(number);
+     }
+
+Stage this hunk [y,n,q,a,d,K,g,/,s,e,?]? y
+```
+
+I have to verify also if there is something else to add that has not been added:
+
+```bash
+[jaar@port-staff api]$ git status
+On branch bugsnag-integration
+Changes to be committed:
+  (use "git reset HEAD <file>..." to unstage)
+
+        modified:   build.gradle
+        modified:   src/main/java/co/org/osso/api/CalculatorController.java
+
+Untracked files:
+  (use "git add <file>..." to include in what will be committed)
+
+        src/main/java/co/org/osso/api/BugsnagConfig.java
+
+[jaar@port-staff api]$ git add .
+```
+
+#### Debugging example
+
+Instead to be printing lot of messages to console its more useful and flexible to debug the code, this is done adding breakpoints (red circles) to the code and once the application is stopped, it can be debbuged clicking on the bug icon or mouse right click menu and 'Debug Application'.  
```

Comment:
FYI: You can also debug tests - which quickly becomes more useful than debugging an application as your application will get very big and re-creating a bug isn't always going to remain this simple.